File: src/truck/truck.py
"""Tractor-trailer mobile robot simulation (standalone, matplotlib-based).

Mirrors Truck.js: same state, same controls, same skeleton left to fill in.

State:   x_c, y_c   (hitch position — where cabin and trailer connect)
         theta_c    (cabin heading, radians, CCW from +x axis)
         theta_t    (trailer heading, radians, CCW from +x axis)
Control: v (cabin linear velocity), phi (cabin angular velocity)

Configurable geometry:
    L (L)          cabin is a square, hitch sits at its rear,
                              cabin body extends forward from the hitch by L.
    d (d) trailer body extends backward from the hitch
                              to its rear axle, a distance d away.

Controls (while the plot window has focus):
    w   accelerate forward (increases v)
    a   turn CCW (increases phi)
    d   turn CW (decreases phi)

Requires matplotlib + numpy (not in requirements.txt — this is a standalone
dev tool, not part of the Flask backend).
"""

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from matplotlib.animation import FuncAnimation
from matplotlib.transforms import Affine2D

L = 0.4  # L, configurable — cabin is L x L (square)
d = 0.8  # d, configurable — hitch-to-rear-axle coupling distance
L2 = 0.7 # L2, configurable — fixed trailer body length (must be <= d)
W = 0.4  # units, across trailer heading

LINEAR_ACCEL = 1.5  # units/s^2 while "w" is held
LINEAR_DAMPING = 2.0  # 1/s, decays v back to 0 once "w" is released
MAX_LINEAR_VEL = 1.5  # units/s

ANGULAR_ACCEL = 10.0 * np.pi / 180  # rad/s^2 while "a"/"d" is held
ANGULAR_DAMPING = 3.0  # 1/s, decays phi back to 0 once "a"/"d" is released
MAX_STEERING_PHI = 12 * np.pi / 180

DT = 1 / 30.0


class Truck:
    def __init__(
        self,
        x_c=0.0,
        y_c=0.0,
        theta_c=0.0,
        theta_t=0.0,
        x_t=None,
        y_t=None,
    ):
        self.x_c = x_c
        self.y_c = y_c
        self.theta_c = theta_c
        self.theta_t = theta_t
        # Rear axle of the trailer — tracked explicitly (not derived from
        # x_c/y_c/theta_t/d each frame). Defaults to directly behind the
        # hitch if not given.
        self.x_t = x_c - d * np.cos(theta_t)
        self.y_t = y_c - d * np.sin(theta_t)
        self.v = 0.0
        self.phi = 0.0

    def apply_controls(self, keys, dt):
        """Maps currently-held keys -> (v, phi) via accelerate/damping.

        This is control input, not the kinematics, so it's fully wired up
        already — see step() below for what's left to fill in.
        """
        if keys.get("w"):
            self.v = min(MAX_LINEAR_VEL, self.v + LINEAR_ACCEL * dt)
        elif self.v > 0:
            self.v = max(0.0, self.v - LINEAR_DAMPING * MAX_LINEAR_VEL * dt)

        # phi is the steering angle *relative to theta_c* (angle between the
        # front wheel's perpendicular and the cabin heading) — step() uses
        # tan(phi) directly, so phi itself must stay within a fixed band
        # around 0. (It must NOT track theta_c: doing so turns this into a
        # feedback loop, since a bigger phi speeds up theta_c, which would
        # then keep raising phi's own ceiling.)
        turning_ccw = keys.get("a") and not keys.get("d")
        turning_cw = keys.get("d") and not keys.get("a")
        if turning_ccw:
            self.phi = min(MAX_STEERING_PHI, self.phi + ANGULAR_ACCEL * dt)
        elif turning_cw:
            self.phi = max(-MAX_STEERING_PHI, self.phi - ANGULAR_ACCEL * dt)
        # else:
        #     decay = ANGULAR_DAMPING * dt
        #     if self.phi > 0:
        #         self.phi = max(0.0, self.phi - decay)
        #     elif self.phi < 0:
        #         self.phi = min(0.0, self.phi + decay)

    def step(self, dt):
        dx_c = self.v * np.cos(self.theta_c)
        dy_c = self.v * np.sin(self.theta_c)
        dtheta_c = self.v / (L / 2) * np.tan(self.phi)
        dtheta_t = self.v / d * np.sin(self.theta_c - self.theta_t)

        dx_t = dx_c + d * dtheta_t * np.sin(self.theta_t)
        dy_t = dy_c - d * dtheta_t * np.cos(self.theta_t)

        self.x_c += dx_c * dt
        self.y_c += dy_c * dt
        self.theta_c += dtheta_c * dt
        self.theta_t += dtheta_t * dt
        self.x_t += dx_t * dt
        self.y_t += dy_t * dt


def main():
    truck = Truck()
    keys = {"w": False, "a": False, "d": False}

    fig, ax = plt.subplots(figsize=(7, 7))
    ax.set_xlim(-10, 10)
    ax.set_ylim(-10, 10)
    ax.set_aspect("equal")
    ax.set_title("Tractor-trailer robot  (W: accelerate, A/D: turn)")
    ax.grid(True, linewidth=0.3)

    # Cabin: square, extends forward from the hitch along theta_c.
    cabin = patches.Rectangle(
        (0, -L / 2),
        L,
        L,
        facecolor="#4da3ff",
        edgecolor="black",
    )
    ax.add_patch(cabin)

    # Trailer: fixed length L2 (<= d), rear axle at (x_t, y_t), body extends
    # forward from the axle along theta_t.
    trailer = patches.Rectangle(
        (0, -W / 2),
        L2,
        W,
        facecolor="#c97b3d",
        edgecolor="black",
    )
    ax.add_patch(trailer)

    (heading_line,) = ax.plot([], [], color="white", linewidth=2)
    # Tow-bar: connects the hitch to the tracked rear-axle point (x_t, y_t).
    (tow_bar,) = ax.plot([], [], color="#8a8a8a", linewidth=2)

    # Rear-wheel axles: a bar perpendicular to the body's heading, with a
    # wheel dot at each end — for sanity checking that a body's
    # position/heading/rear-axle state line up visually.
    (cabin_axle,) = ax.plot([], [], color="#39ff14", linewidth=3, marker="o", markersize=5)
    (trailer_axle,) = ax.plot([], [], color="#ffe14d", linewidth=3, marker="o", markersize=5)

    def axle_endpoints(x, y, theta, track):
        half_track = track / 2
        perp = theta + np.pi / 2
        dx = np.cos(perp) * half_track
        dy = np.sin(perp) * half_track
        return [x - dx, x + dx], [y - dy, y + dy]
    hud = ax.text(0.02, 0.98, "", transform=ax.transAxes, va="top", family="monospace")

    def on_key(event, pressed):
        if event.key in keys:
            keys[event.key] = pressed

    fig.canvas.mpl_connect("key_press_event", lambda e: on_key(e, True))
    fig.canvas.mpl_connect("key_release_event", lambda e: on_key(e, False))

    def update(_frame):
        truck.apply_controls(keys, DT)
        truck.step(DT)

        cabin_transform = (
            Affine2D().rotate(truck.theta_c).translate(truck.x_c, truck.y_c) + ax.transData
        )
        cabin.set_transform(cabin_transform)

        trailer_transform = (
            Affine2D().rotate(truck.theta_t).translate(truck.x_t, truck.y_t) + ax.transData
        )
        trailer.set_transform(trailer_transform)

        heading_line.set_data(
            [truck.x_c, truck.x_c + L * np.cos(truck.theta_c)],
            [truck.y_c, truck.y_c + L * np.sin(truck.theta_c)],
        )

        tow_bar.set_data([truck.x_c, truck.x_t], [truck.y_c, truck.y_t])

        # cabin rear axle coincides with the hitch, (x_c, y_c), in this model
        cabin_axle.set_data(*axle_endpoints(
            truck.x_c + (L / 2) * np.cos(truck.theta_c), 
            truck.y_c + (L / 2) * np.sin(truck.theta_c), truck.phi, L))
        trailer_axle.set_data(*axle_endpoints(truck.x_t, truck.y_t, truck.theta_t, W))

        hud.set_text(
            f"x_c={truck.x_c:6.2f}  y_c={truck.y_c:6.2f}\n"
            f"theta_c={truck.theta_c:5.2f}  theta_t={truck.theta_t:5.2f}\n"
            f"x_t={truck.x_t:6.2f}  y_t={truck.y_t:6.2f}\n"
            f"v={truck.v:5.2f}  phi={truck.phi:5.2f}"
        )

        return cabin, trailer, heading_line, tow_bar, cabin_axle, trailer_axle, hud

    _anim = FuncAnimation(fig, update, interval=DT * 1000, blit=False)
    plt.show()


if __name__ == "__main__":
    main()

File: src/truck/Truck.js
import { useCallback, useEffect, useRef, useState, useLayoutEffect } from "react";
import { generateTrajectoryPNG } from "./Plot.js";
import Draggable from "react-draggable";
import App from '../App';

const L = 40; // L : square cabin length
const d = 80; // d : dist from rear axle to hitch
const L2 = 70; // L2 : trailer length
const W = 40; // W : same as cabin length 

const LINEAR_ACCEL = 100; // px/s^2 while "w" is held
const LINEAR_DAMPING = 2.0; // 1/s, decays v back to 0 once "w" is released
const MAX_LINEAR_VEL = 200; // px/s

const ANGULAR_ACCEL = 10 * Math.PI / 180; // acceleration of the steering angle
// const ANGULAR_DAMPING = 3.0;
const MAX_STEERING_PHI = 12 * Math.PI / 180; // max steering angle

const CANVAS_WIDTH = 800;
const CANVAS_HEIGHT = 600;

// Initial state = [xc, yc, thetac, thetat].T
const INITIAL_X_C = 600;
const INITIAL_Y_C = 400;
const INITIAL_THETA_C = -80 * Math.PI / 180;
const INITIAL_THETA_T = -70 * Math.PI / 180;

// Rear axle
const INITIAL_X_T = INITIAL_X_C - d * Math.cos(INITIAL_THETA_T);
const INITIAL_Y_T = INITIAL_Y_C - d * Math.sin(INITIAL_THETA_T);

function stepKinematics(state, dt) {
  const { x_c, y_c, theta_c, theta_t, x_t, y_t, v, phi } = state;

  const dx_c = v * Math.cos(theta_c);
  const dy_c = v * Math.sin(theta_c);
  const dtheta_c = v / (L / 2) * Math.tan(phi);
  const dtheta_t = v / d * Math.sin(theta_c - theta_t);
  const dx_t = dx_c + d * dtheta_t * Math.sin(theta_t);
  const dy_t = dy_c - d * dtheta_t * Math.cos(theta_t);

  return {
    ...state,
    x_c: x_c + dx_c * dt,
    y_c: y_c + dy_c * dt,
    theta_c: theta_c + dtheta_c * dt,
    theta_t: theta_t + dtheta_t * dt,
    x_t: x_t + dx_t * dt,
    y_t: y_t + dy_t * dt,
  };
}

function applyControls(state, keys, dt) {
  let { v, phi } = state;

  if (keys.w) {
    v = Math.min(MAX_LINEAR_VEL, v + LINEAR_ACCEL * dt);
  } else if (keys.s) {
    v = Math.max(-MAX_LINEAR_VEL, v - LINEAR_ACCEL * dt);
  } else if (Math.abs(v) > 0) {
    let sign = (v > 0) ? 1 : -1;
    v = sign * Math.max(0, Math.abs(v) - LINEAR_DAMPING * MAX_LINEAR_VEL * dt);
  }

  // phi is the steering angle: the angle between perpendicular bisector of cabin axle 
  // and heading or orientation of the cabin
  const turningCCW = keys.a && !keys.d;
  const turningCW = keys.d && !keys.a;
  if (turningCCW) {
    phi = Math.min(MAX_STEERING_PHI, phi + ANGULAR_ACCEL * dt);
  } else if (turningCW) {
    phi = Math.max(-MAX_STEERING_PHI, phi - ANGULAR_ACCEL * dt);
  }
//   else {
//     const decay = ANGULAR_DAMPING * dt;
//     if (phi > 0) phi = Math.max(0, phi - decay);
//     else if (phi < 0) phi = Math.min(0, phi + decay);
//   }

  return { ...state, v, phi };
}

// draw the cabin and rear axles to visualize how axle orientation changes with phi
function drawAxle(ctx, x, y, theta, track, color) {
  const halfTrack = track / 2;
  const perp = theta + Math.PI / 2;
  const dx = Math.cos(perp) * halfTrack;
  const dy = Math.sin(perp) * halfTrack;

  ctx.save();
  ctx.strokeStyle = color;
  ctx.fillStyle = color;
  ctx.lineWidth = 3;

  ctx.beginPath();
  ctx.moveTo(x - dx, y - dy);
  ctx.lineTo(x + dx, y + dy);
  ctx.stroke();

  ctx.beginPath();
  ctx.arc(x - dx, y - dy, 4, 0, Math.PI * 2);
  ctx.arc(x + dx, y + dy, 4, 0, Math.PI * 2);
  ctx.fill();

  ctx.restore();
}

function drawScene(ctx, state) {
  const { x_c, y_c, theta_c, theta_t, x_t, y_t, v, phi } = state;
  const x_axle = x_c + (L / 2) * Math.cos(theta_c);
  const y_axle = y_c + (L / 2) * Math.sin(theta_c);

  ctx.clearRect(0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);

  ctx.save();
  ctx.translate(0, CANVAS_HEIGHT);
  ctx.scale(1, -1);

  // faint reference grid
  ctx.save();
  ctx.strokeStyle = "#2a2a2a";
  ctx.lineWidth = 1;
  const gridSize = 40;
  for (let gx = 0; gx < CANVAS_WIDTH; gx += gridSize) {
    ctx.beginPath();
    ctx.moveTo(gx, 0);
    ctx.lineTo(gx, CANVAS_HEIGHT);
    ctx.stroke();
  }
  for (let gy = 0; gy < CANVAS_HEIGHT; gy += gridSize) {
    ctx.beginPath();
    ctx.moveTo(0, gy);
    ctx.lineTo(CANVAS_WIDTH, gy);
    ctx.stroke();
  }
  ctx.restore();

  // angle from x-axis to axle: phi + theta_c
  let radius = L / (2 * Math.tan(phi));
  ctx.save();
  ctx.beginPath();
  ctx.arc(
    x_axle - Math.sin(phi + theta_c) * radius, 
    y_axle + Math.cos(phi + theta_c) * radius, 
    Math.abs(radius), 0, Math.PI * 2); // this is to draw the circular path that the cabin is tracing when turn
  ctx.strokeStyle = "#ffffff";
  ctx.lineWidth = 2;
  ctx.stroke();
  ctx.restore();

  // the line connecting rear axle midpoint to hitch, parallel to trailer length, in figure 2.38
  ctx.save();
  ctx.strokeStyle = "#8a8a8a";
  ctx.lineWidth = 2;
  ctx.beginPath();
  ctx.moveTo(x_c, y_c);
  ctx.lineTo(x_t, y_t);
  ctx.stroke();
  ctx.restore();

  // draw the trailer
  ctx.save();
  ctx.translate(x_t, y_t);
  ctx.rotate(theta_t);
  ctx.fillStyle = "#c97b3d";
  ctx.strokeStyle = "#ffe3c9";
  ctx.lineWidth = 2;
  ctx.fillRect(0, -W / 2, L2, W);
  ctx.strokeRect(0, -W / 2, L2, W);
  ctx.restore();

  // draw the trailer rear axle
  drawAxle(ctx, x_t, y_t, theta_t, W, "#ffe14d");

  // draw the cabin
  ctx.save();
  ctx.translate(x_c, y_c);
  ctx.rotate(theta_c);
  ctx.fillStyle = "#4da3ff";
  ctx.strokeStyle = "#dff0ff";
  ctx.lineWidth = 2;
  ctx.fillRect(0, -L / 2, L, L);
  ctx.strokeRect(0, -L / 2, L, L);

  // depict front of the cabin
  ctx.fillStyle = "#ffffff";
  ctx.beginPath();
  ctx.moveTo(L, 0);
  ctx.lineTo(L - 10, -8);
  ctx.lineTo(L - 10, 8);
  ctx.closePath();
  ctx.fill();
  ctx.restore();

  // draw the cabin axle
  drawAxle(ctx, 
    x_axle, y_axle, phi + theta_c, L, "#39ff14");
 
  ctx.restore();
}

// models the geometry and kinematics of the semi-trailer truck
export default function Truck() {
  // stores states and controls so i can plot truck path across time period
  const screenWidth = 84;
  const [width, setWidth] = useState(0);
  const [height, setHeight] = useState(0);
//   const trajectoryRef = useRef([{
//         t: 0,
//         x_c: INITIAL_X_C,
//         y_c: INITIAL_Y_C,
//         theta_c: INITIAL_THETA_C,
//         x_t: INITIAL_X_T,
//         y_t: INITIAL_Y_T,
//         theta_t: INITIAL_THETA_T,
//         v: 0,
//         phi: 0,
//     }]);
//   const elapsedRef = useRef(0);
//   const nextPlotTimeRef = useRef(0);
  const containerRef = useRef(null);
  const canvasRef = useRef(null);
  const keysRef = useRef({ w: false, a: false, d: false });
  const stateRef = useRef({
    x_c: INITIAL_X_C,
    y_c: INITIAL_Y_C,
    theta_c: INITIAL_THETA_C,
    theta_t: INITIAL_THETA_T,
    x_t: INITIAL_X_T,
    y_t: INITIAL_Y_T,
    v: 0,
    phi: 0,
  });
  const [hud, setHud] = useState(stateRef.current);

  const handleKeyDown = useCallback((event) => {
    const key = event.key.toLowerCase();
    if (key === "w" || key === "a" || key === "d" || key === "s") {
      keysRef.current[key] = true;
    }
  }, []);

  const handleKeyUp = useCallback((event) => {
    const key = event.key.toLowerCase();
    if (key === "w" || key === "a" || key === "d" || key === "s") {
      keysRef.current[key] = false;
    }
  }, []);

  useLayoutEffect(() => {
    const updateWidth = () => {
        if (containerRef.current) {
            setWidth(containerRef.current.clientWidth);
            setHeight(containerRef.current.clientHeight);
        }
    };

    updateWidth();

    const resizeObserver = new ResizeObserver(updateWidth);
    if (containerRef.current) {
        resizeObserver.observe(containerRef.current);
    }

    window.addEventListener('resize', updateWidth);

    return () => {
        resizeObserver.disconnect();
        window.removeEventListener('resize', updateWidth);
    };
}, []);

  useEffect(() => {
    window.addEventListener("keydown", handleKeyDown);
    window.addEventListener("keyup", handleKeyUp);
    return () => {
      window.removeEventListener("keydown", handleKeyDown);
      window.removeEventListener("keyup", handleKeyUp);
    };
  }, [handleKeyDown, handleKeyUp]);

  useEffect(() => {
    const ctx = canvasRef.current.getContext("2d");
    let frameId;
    let lastTime = performance.now();

    const tick = (now) => {
      const dt = Math.min(0.05, (now - lastTime) / 1000);
      lastTime = now;

      let state = applyControls(stateRef.current, keysRef.current, dt);
      state = stepKinematics(state, dt);
      stateRef.current = state;

    //   elapsedRef.current += dt;

    //   if (elapsedRef.current >= nextPlotTimeRef.current) {
    //     trajectoryRef.current.push({
    //         t: nextPlotTimeRef.current,
    //         x_c: state.x_c,
    //         y_c: state.y_c,
    //         theta_c: state.theta_c,
    //         x_t: state.x_t,
    //         y_t: state.y_t,
    //         theta_t: state.theta_t,
    //         v: state.v,
    //         phi: state.phi * 180 / Math.PI
    //     });

    //     nextPlotTimeRef.current += 0.5;

        // stop recording after t = 10s
    //     if (nextPlotTimeRef.current > 10 && trajectoryRef.current != []) {
    //         generateTrajectoryPNG(trajectoryRef);

    //         trajectoryRef.current = [];
    //         nextPlotTimeRef.current = Infinity;
    //     }
    //   }

      drawScene(ctx, state);
      setHud(state);

      frameId = requestAnimationFrame(tick);
    };

    frameId = requestAnimationFrame(tick);
    return () => cancelAnimationFrame(frameId);
  }, []);

  return (
    <App enableUI={false} responsivePopup popup={
        <Draggable
            cancel={".draggable-btn"}
            nodeRef={containerRef}
            bounds={{ left: 0, top: 0, right: width + 140, bottom: height - 500}}
        >  
        <div ref={containerRef} style={{ 
            width: 1000,
            padding: 2, 
            fontFamily: 'Pixelify Sans',
            position: 'relative',
            display: 'inline-block'
        }}>
            <img 
                className="lidar_window"
                src={process.env.PUBLIC_URL + "/assets/window.png"} 
                style={{
                    zIndex: 0,
                    position: 'absolute',
                    left: (screenWidth / 2 - 67 / 2) * width / screenWidth,
                    width: width * 67 / screenWidth,
                    display: 'block'
                }}
            />

            <canvas
                ref={canvasRef}
                width={width * 57 / screenWidth}
                height={335}
                style={{ 
                    position: "relative", 
                    display: "block",
                    top: 9 * width / screenWidth,
                    left: (screenWidth / 2 - 57.3 / 2) * width / screenWidth,
                    background: "#1b1b1b" }}
            />
            
        
            {/* <div
                style={{
                position: "absolute",
                top: 12,
                left: 12,
                color: "#fff",
                fontFamily: "monospace",
                fontSize: 13,
                lineHeight: 1.6,
                background: "rgba(0,0,0,0.5)",
                padding: "8px 12px",
                borderRadius: 6,
                }}
            >
                <div>W: accelerate&nbsp;&nbsp; A: turn CCW&nbsp;&nbsp; D: turn CW</div>
                <div>
                x_c: {hud.x_c.toFixed(1)}&nbsp;&nbsp; y_c: {hud.y_c.toFixed(1)}
                </div>
                <div>
                theta_c: {hud.theta_c.toFixed(2)} rad&nbsp;&nbsp; theta_t: {hud.theta_t.toFixed(2)} rad
                </div>
                <div>
                x_t: {hud.x_t.toFixed(1)}&nbsp;&nbsp; y_t: {hud.y_t.toFixed(1)}
                </div>
                <div>
                v: {hud.v.toFixed(1)}&nbsp;&nbsp; phi: {hud.phi.toFixed(2)}
                </div>
            </div> */}
        </div>
        </Draggable>}
        animate={false}
    />
  );
}
